src/providers/activeLearning/objectDetection.ts

```diff
@@ -0,0 +1,197 @@
+            }
+            // Reshape to a single-element batch so we can pass it to executeAsync.
+            return img.expandDims(0);
+        });
+        const height = batched.shape[1];
+        const width = batched.shape[2];
+
+        // model returns two tensors:
+        // 1. box classification score with shape of [1, 1917, 90]
+        // 2. box location with shape of [1, 1917, 1, 4]
+        // where 1917 is the number of box detectors, 90 is the number of classes.
+        // and 4 is the four coordinates of the box.
+        const result = await this.model.executeAsync(batched) as tf.Tensor[];
+
+        const scores = result[0].dataSync() as Float32Array;
+        const boxes = result[1].dataSync() as Float32Array;
+
+        // clean the webgl tensors
+        batched.dispose();
+        tf.dispose(result);
+
+        const [maxScores, classes] = this.calculateMaxScores(scores, result[0].shape[1], result[0].shape[2]);
+
+        const prevBackend = tf.getBackend();
+        // run post process in cpu
+        tf.setBackend("cpu");
+        const indexTensor = tf.tidy(() => {
+            const boxes2 =
+                tf.tensor2d(boxes, [result[1].shape[1], result[1].shape[3]]);
+            return tf.image.nonMaxSuppression(
+                boxes2, maxScores, maxNumBoxes, 0.5, 0.5);
+        });
+
+        const indexes = indexTensor.dataSync() as Float32Array;
+        indexTensor.dispose();
+
+        // restore previous backend
+        tf.setBackend(prevBackend);
+
+        return this.buildDetectedObjects(width, height, boxes, maxScores, indexes, classes);
+    }
+
+    private buildDetectedObjects(
+        width: number, height: number, boxes: Float32Array, scores: number[],
+        indexes: Float32Array, classes: number[]): DetectedObject[] {
+        const count = indexes.length;
+        const objects: DetectedObject[] = [];
+        for (let i = 0; i < count; i++) {
+            const bbox = [];
+            for (let j = 0; j < 4; j++) {
+                bbox[j] = boxes[indexes[i] * 4 + j];
+            }
+            const minY = bbox[0] * height;
+            const minX = bbox[1] * width;
+            const maxY = bbox[2] * height;
+            const maxX = bbox[3] * width;
+            bbox[0] = minX;
+            bbox[1] = minY;
+            bbox[2] = maxX - minX;
+            bbox[3] = maxY - minY;
+            objects.push({
+                bbox: bbox as [number, number, number, number],
+                class: this.getClass(i, indexes, classes),
+                score: scores[indexes[i]],
+            });
+        }
+        return objects;
+    }
+
+    private getClass(index: number, indexes: Float32Array, classes: number[]): string {
+        if (index < indexes.length && indexes[index] < classes.length) {
+            const classId = classes[indexes[index]] - 1;
+            const classObject = this.jsonClasses[classId];
+            return classObject ? classObject.displayName : "";
```

Comment:
would 'unknown' or 'not found' be more informative than an empty string? Is this displayed in the UI? Trying to think of how it would hit this case...

Reply:
Good point.  Yes this is also UI related.  It is the tag name to use just in case there is a incongruence between the model and the json file with all tags managed by the model.  The orginal Google code for Coco image detection was simply embedding all this tags in code in a js object.  I extended this with this concept of a json file to dinamically pass a list of all the tags your model detect.  This is because of course the model doesn't use strings but it encode that in categorized label (numbers).  

Anyway probably you're right and "Unknown" is a much better label than an empty one!!!